Opt into unsafe lifecycle warnings without async tree

bvaughn opened the pull request:
Opts a subtree into the same warnings as `AsyncComponent` does, but does not actually enable async rendering. This same mechanism could be extended for similar warnings in the future (eg compiler, React Fabric).

Following precedent, this component is exposed via `React.StrictMode`. It is used like a `Fragment` to wrap an element:

```js
const {StrictMode} = React;

<StrictMode>
  {/* Subtree here */}
</StrictMode>
```

This branch builds on top of #12060.

sebmarkbage commented on packages/react/src/ReactBaseClasses.js:
```diff
@@ -117,44 +117,56 @@ if (__DEV__) {
-const pureComponentPrototype = (PureComponent.prototype = new ComponentDummy());
-pureComponentPrototype.constructor = PureComponent;
-// Avoid an extra prototype jump for these methods.
-Object.assign(pureComponentPrototype, Component.prototype);
```
Are we sure this matters these days in any significant way? Maybe we can just kill this.

sebmarkbage commented:
@acdlite Originally we thought that we needed to be able to inherit AsyncComponent instead of wrapping, because at the top level it is tricky since it affects the return value of render. However, if you're upgrading at the top level, you probably want to use the createRoot API instead. Do we still think that we want the inheritance mechanism or should we switch to what context and fragment do?

bvaughn commented:
> Duplication is not a good enough reason to introduce an abstraction.

The code felt crufty and repetitive before, so I proposed a change. I'll revert it.

> Maybe we can rename it to something like `<EnableFutureCompatibilityChecks />` or something? I don't know.

Interesting. I like this suggestion. Not sure about _that name_ but I like the suggestion.

gaearon commented:
To give some background on the abstraction thing, the code feeling crufty/duplicated is how we ended up with a lot of things in the Stack implementation (like "Transactions") that ended up being hard to work with. So these days we try to err on the side of making *removal* of the code easy rather than making it feel dry.

You know why the abstraction was added and thus know when to remove it. But other people often don’t, and as time goes everybody is afraid to rip out each other’s abstractions, and we end up with overabstracted code that *seems* to be necessary but no one’s really sure. Hope this makes sense 😛

bvaughn commented:
Sure, it makes sense. I didn't feel strongly about the change anyway. I mentioned in the initial PR description that I would be happy to revert it if there were any concerns, for this reason ^

bvaughn commented:
Based on discussion above and in person (during lunch) I've updated this PR as follows:
* The new component type is named `StrictMode` (rather than anything to do with async).
* The new internal context flag is also named `StrictMode`.
* Async subtrees are also `StrictMode` subtrees and all of the previous async warning checks are now based on this flag instead.
* The `ReactDebugAsyncWarnings` helper has been renamed `ReactStrictModeWarnings` to avoid confusion.
* A warning message has been updated with a new fb.me link, https://fb.me/react-strict-mode-warnings

sebmarkbage commented:
What did you think of @tomocchino's `ReactDOM.renderStrict(...)` proposal to opt in a whole tree?

That would eliminate the need for base class based opt in `class Foo extends StrictMode` which seems weird.

bvaughn commented:
> What did you think of @tomocchino's `ReactDOM.renderStrict(...)` proposal to opt in a whole tree?

I think I like it as a follow-up option, but it doesn't seem sufficient to replace `React.StrictComponent`.

sebmarkbage commented:
If you're testing your children you can always wrap `return <StrictMode>{myChildren}</StrictMode>`. The only thing it helps is with testing yourself but I'm not sure how valuable that is.

bvaughn commented:
Oh I think I misunderstood your question!

bvaughn commented:
Okay! I converted `StrictMode` to a symbol, verified all previous tests passed, and added a couple of new tests to cover going between `StrictMode` -> `Fragment`, `Fragment` -> `StrictMode`, and `StrictMode` -> `StrictMode`.

sebmarkbage commented on packages/shared/ReactSymbols.js:
```diff
@@ -24,6 +24,9 @@ export const REACT_PORTAL_TYPE = hasSymbol
   ? Symbol.for('react.fragment')
   : 0xeacb;
+export const REACT_USE_STRICT_TYPE = hasSymbol
+  ? Symbol.for('react.use_strict')
```
Let's call it the same thing as what we export. `react.strict_mode`

`use strict` is JavaScript.

gaearon replied:
But ... React is JavaScript! 😄

sebmarkbage commented on packages/react-reconciler/src/ReactFiberClassComponent.js:
```diff
@@ -639,20 +639,23 @@ export default function(
+        }
+      }
+
+      if (prototype.__reactStrictMode === true) {
```
This seems like it is not necessary anymore.

sebmarkbage commented on packages/react/src/ReactBaseClasses.js:
```diff
@@ -153,8 +151,30 @@ asyncComponentPrototype.constructor = AsyncComponent;
+ * For examples, it detects async-unsafe lifecycles (without actually enabling async mode).
+ * It can also detect render-phase side effects by double-invoking certain methods.
+ */
+function StrictMode(props, context, updater) {
```
We can remove this now.

bvaughn replied:
Yeah, this was the silly oversight ;)

bvaughn commented:
Whoops. I made a silly oversight. Give me a moment.

gaearon commented:
I *love* the strict mode parallel. This is a much easier story to convey.

bvaughn commented:
Okay. A new type-of-work `Mode` has been added.
